File: routes/complaints.ts
// Complaints Routes
import { Express } from "express";

export function registerComplaintRoutes(app: Express, db: any) {
  app.get("/api/complaints", (req, res) => {
    try {
      res.json({ success: true, data: db.prepare("SELECT * FROM complaints ORDER BY id DESC").all() });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  app.get("/api/complaints/my/:userId", (req, res) => {
    try {
      res.json({ success: true, data: db.prepare("SELECT * FROM complaints WHERE user_id = ?").all(parseInt(req.params.userId)) });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  app.post("/api/complaints", (req, res) => {
    try {
      const { user_id, category, subject, description } = req.body;
      const today = new Date().toISOString().split('T')[0];
      db.prepare("INSERT INTO complaints (user_id, date, category, subject, description) VALUES (?, ?, ?, ?, ?)").run(user_id, today, category, subject, description);
      res.json({ success: true });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  app.put("/api/complaints/:id/respond", (req, res) => {
    try {
      const { response, status } = req.body;
      db.prepare("UPDATE complaints SET response=?, status=? WHERE id=?").run(response, status || 'resolved', parseInt(req.params.id));
      res.json({ success: true });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  app.put("/api/complaints/:id/status", (req, res) => {
    try {
      db.prepare("UPDATE complaints SET status=? WHERE id=?").run(req.body.status, parseInt(req.params.id));
      res.json({ success: true });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });
}

File: backend/routes/complaints.py
from fastapi import APIRouter
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
from database import db

router = APIRouter(prefix="/api", tags=["complaints"])


class ComplaintCreate(BaseModel):
    user_id: int
    category: str
    subject: str
    description: str


class ComplaintRespond(BaseModel):
    response: str
    status: Optional[str] = "resolved"


class StatusUpdate(BaseModel):
    status: str


@router.get("/complaints")
def list_complaints():
    rows = db.get_all("complaints")
    return {"success": True, "data": sorted(rows, key=lambda c: c.get("id", 0), reverse=True)}


@router.get("/complaints/my/{user_id}")
def my_complaints(user_id: int):
    return {"success": True, "data": db.filter("complaints", user_id=user_id)}


@router.post("/complaints")
def create_complaint(c: ComplaintCreate):
    today = datetime.now().strftime("%Y-%m-%d")
    db.insert("complaints", {
        "user_id": c.user_id, "date": today, "category": c.category,
        "subject": c.subject, "description": c.description,
        "status": "pending", "response": None,
        "responded_by": None, "responded_at": None,
        "created_at": datetime.now().isoformat(),
    })
    return {"success": True}


@router.put("/complaints/{cid}/respond")
def respond_complaint(cid: int, body: ComplaintRespond):
    db.update("complaints", cid, {
        "response": body.response,
        "status": body.status or "resolved",
        "responded_at": datetime.now().isoformat(),
    })
    return {"success": True}


@router.put("/complaints/{cid}/status")
def update_complaint_status(cid: int, body: StatusUpdate):
    db.update("complaints", cid, {"status": body.status})
    return {"success": True}
